File: scripts/clawtributor-campaign50/lanes/prepared-catalog-green-lint-remaining/validate-reuse.mjs
import assert from "node:assert/strict";
import { spawnSync } from "node:child_process";
import { createHash } from "node:crypto";
import fs from "node:fs";
import path from "node:path";

const [laneDir, evidenceDir] = process.argv.slice(2);
const reuseDir = path.join(laneDir, "reuse");
const readJson = (file) => JSON.parse(fs.readFileSync(path.join(reuseDir, file), "utf8"));
const lineage = readJson("LINEAGE.json");
assert.equal(lineage.run, 34067314611);
assert.equal(lineage.job, 101578264433);
assert.equal(lineage.harness, "f35646704bdfafd0e8a5368c11c945f1dd59b8f8");
assert.equal(lineage.source, "f63095668169ea0260f5f7b6479dafc85cc5226c");
assert.equal(lineage.candidate, "1a42e4917dfda234bd3a4e32dfcd365a06baf4cf3271497fbda544fcc6f3704b");
assert.equal(
  lineage.correctedReader,
  "01ee2284d0bdd3f4f3120db0b9f63d13b808eb38e1b9f3b8380a1429a1b8c03e",
);
const expectedArtifacts = [
  "unit.json",
  "unit.log.txt",
  "source.json",
  "unchanged-source.sha256",
  "final-working-tree.patch",
  "index.json",
  "aggregate.capture.json",
  "merge-config.mjs.txt",
  "run.json",
  "memory-job.log.txt",
  ...[1, 2, 3].flatMap((index) => [`shard-${index}.json`, `shard-${index}.capture.json`]),
];
assert.deepEqual(
  lineage.artifacts.map((entry) => entry.transportPath).sort(),
  expectedArtifacts.sort(),
);
for (const entry of lineage.artifacts) {
  const data = fs.readFileSync(path.join(reuseDir, entry.transportPath));
  assert.equal(data.length, entry.bytes, entry.transportPath);
  assert.equal(createHash("sha256").update(data).digest("hex"), entry.sha256, entry.transportPath);
}
const source = readJson("source.json");
assert.equal(source.source, lineage.source);
assert.equal(source.lane, "prepared-catalog-green");
assert.equal(source.mode, "green");
assert.equal(source.node, "24.20.0");
assert.ok(source.packageManager.startsWith("pnpm@12.3.4+"));
const run = readJson("run.json");
assert.equal(run.databaseId, lineage.run);
assert.equal(run.headSha, lineage.harness);
const jobs = run.jobs.filter((job) => job.databaseId === lineage.job);
assert.equal(jobs.length, 1);
assert.equal(jobs[0].status, "completed");
assert.equal(jobs[0].conclusion, "failure");

const outcome = { code: 0, noOutputTimedOut: false, signal: null, groupJoined: true };
const index = readJson("index.json");
assert.equal(index.complete, true);
assert.equal(index.error, "");
assert.equal(index.entries.length, 3);
assert.deepEqual(index.merge, outcome);
const rawTests = [];
const rawFiles = [];
const rawCaptures = [];
const normalizedTests = (report) =>
  report.testResults
    .flatMap((suite) =>
      suite.assertionResults.map((test) => ({
        file: path.basename(suite.name),
        fullName: test.fullName,
        status: test.status,
      })),
    )
    .map((test) => JSON.stringify(test))
    .sort();
const assertCapture = (capture, ignoreUnhandledErrors = false) => {
  assert.equal(capture.processTimedOut, false);
  assert.equal(capture.ignoreUnhandledErrors, ignoreUnhandledErrors);
  assert.equal(capture.passWithNoTests, false);
  assert.deepEqual(capture.ended, {
    reason: "passed",
    unhandledErrors: 0,
    failedModules: 0,
    suiteErrors: 0,
  });
};
for (const invocation of [1, 2, 3]) {
  const entries = index.entries.filter((entry) => entry.invocation === invocation);
  assert.equal(entries.length, 1);
  const entry = entries[0];
  assert.equal(entry.state, "finished");
  assert.equal(entry.acceptedAttempt, 1);
  assert.equal(entry.attempts.length, 1);
  assert.deepEqual(entry.attempts[0].outcome, outcome);
  const raw = readJson(`shard-${invocation}.json`);
  assert.equal(raw.success, true);
  assert.equal(raw.numFailedTests, 0);
  assert.equal(raw.numPendingTests, 0);
  assert.equal(raw.numTodoTests ?? 0, 0);
  const capture = readJson(`shard-${invocation}.capture.json`);
  assertCapture(capture);
  rawCaptures.push(capture);
  assert.deepEqual(
    capture.modules.map((module) => path.basename(module.file)).sort(),
    entry.includePatterns.map((file) => path.basename(file)).sort(),
  );
  rawFiles.push(...raw.testResults.map((suite) => path.basename(suite.name)));
  rawTests.push(...normalizedTests(raw));
}
const mergeSource = fs.readFileSync(path.join(reuseDir, "merge-config.mjs.txt"), "utf8");
assert.ok(mergeSource.startsWith("export default ") && mergeSource.endsWith(";\n"));
const mergeConfig = JSON.parse(mergeSource.slice("export default ".length, -2));
assert.equal(mergeConfig.test.passWithNoTests, false);
assert.equal(mergeConfig.test.dangerouslyIgnoreUnhandledErrors, true);
assert.deepEqual(mergeConfig.test.reporters[1][1].expected, rawCaptures);
const aggregateCapture = readJson("aggregate.capture.json");
assert.ok(aggregateCapture.command.includes("--mergeReports"));
// Native replay derives this flag from already-qualified zero-error input captures.
assertCapture(aggregateCapture, true);
assert.deepEqual(rawFiles.sort(), Object.keys(lineage.unitCounts).sort());
const aggregate = readJson("unit.json");
assert.deepEqual(rawTests.sort(), normalizedTests(aggregate));
assert.equal(rawTests.length, 261);
assert.deepEqual(
  Object.fromEntries(
    aggregate.testResults.map((suite) => [
      path.basename(suite.name),
      suite.assertionResults.length,
    ]),
  ),
  lineage.unitCounts,
);
const readerPath = path.join(laneDir, "validate-tests.mjs");
assert.equal(
  createHash("sha256").update(fs.readFileSync(readerPath)).digest("hex"),
  lineage.correctedReader,
);
const reader = spawnSync(
  process.execPath,
  [readerPath, path.join(reuseDir, "unit.json"), path.join(reuseDir, "unit.log.txt")],
  { encoding: "utf8" },
);
assert.equal(reader.error, undefined);
assert.equal(reader.signal, null);
assert.equal(reader.status, 0, reader.stderr);
assert.equal(reader.stdout.trim(), "PREPARED_CATALOG_GREEN_UNITS_CONFIRMED files=5 tests=261");
const receipt = {
  status: "qualified-reused-unit-green",
  run: lineage.run,
  job: lineage.job,
  harness: lineage.harness,
  source: lineage.source,
  candidate: lineage.candidate,
  files: 5,
  tests: 261,
  firstAttempts: 3,
  targetTestsExecuted: false,
};
fs.mkdirSync(evidenceDir, { recursive: true });
fs.writeFileSync(
  path.join(evidenceDir, "reused-unit-verification.json"),
  `${JSON.stringify(receipt, null, 2)}\n`,
);
console.log("PREPARED_CATALOG_REUSED_261_UNITS_CONFIRMED");

File: scripts/clawtributor-campaign50/lanes/progress-width-green-remaining/validate-reuse.py
import hashlib,json,re,sys
from pathlib import Path

sha=lambda b:hashlib.sha256(b).hexdigest()
strip=lambda t:re.sub(r'\x1b\[[0-?]*[ -/]*[@-~]','',re.sub(r'\x1b\].*?(?:\x07|\x1b\\)','',t,flags=re.S))
PASSED=['conflict markers','max-lines suppression ratchet','assertion SAFETY comment ratchet','changelog attributions','doctor deprecation registry','guarded extension wildcard re-exports','plugin-sdk wildcard re-exports','duplicate scan target coverage','dependency pin guard','format changed files','plugin boundaries','wrapper shadowing','package patch guard','test temp creation report (warning-only)','core tsgo graph boundary','typecheck core','typecheck core tests','coercion helper declaration guard','deprecated API usage','dead export scan (skip with OPENCLAW_CHECK_CHANGED_SKIP_DEADCODE=1)']

def validate(data,lineage,repair,before,after):
 def raw(name):return data[name]
 def text(name):return raw(name).decode('utf-8')
 def rd(name):return json.loads(raw(name))
 assert set(data)==set(lineage['files'])
 for f,entry in lineage['files'].items():assert sha(raw(f))==entry['sha256'],f
 run=rd('run.json');assert run['databaseId']==lineage['run']==34169880359 and run['headSha']==lineage['harness']=='dd26a9d5d23978616e97ef11cda50c149fdc0d91' and run['status']=='completed' and run['conclusion']=='failure'
 job=next(j for j in run['jobs'] if j['databaseId']==lineage['job']);assert job['databaseId']==101888055019 and job['status']=='completed' and job['conclusion']=='failure'
 assert lineage['source']=='6aa09cbadb594c3d46d5bd49a28c514df1b256b0'
 inp=rd('input-before.json');packet=rd('harness/PACKET.json');assert inp['packetSha256']==lineage['packetSha256']==sha(raw('harness/PACKET.json'));assert inp['files']==packet['files']
 for name in ['cases.json','SOURCE-TESTS.json','source.sha256','source-green.sha256','check.mts','child.mts','driver.py','inspect-group.mjs','validate-tests.mjs','completed-report.mjs','run.sh']:assert sha(raw('harness/'+name))==inp['files'][name],name
 src=rd('source.json');assert src['source']==lineage['source'] and src['node']=='24.20.0' and src['lane']=='progress-width-green' and src['mode']=='green' and src['packageManager'].startswith('pnpm@12.3.4+')
 for file,pins in [('source-before.log.txt','harness/source.sha256'),('candidate-before.log.txt','harness/source-green.sha256')]:
  expected=[line.split('  ',1)[1]+': OK' for line in text(pins).splitlines()];assert len(expected)==23 and text(file).splitlines()==expected
 assert text('unit-green-exit.txt').strip()=='0' and text('exit-code.txt').strip()=='1'
 assert text('driver.log.txt')==text('check.log.txt')==''
 assert rd('unit-acceptance.json')=={'accepted':True,'files':2,'passed':50,'failed':0}
 assert rd('acceptance.json')=={'accepted':True,'mode':'green','cases':12,'failures':[]}
 unit=rd('unit-green.json');assert unit['success'] and unit['numTotalTests']==unit['numPassedTests']==50
 for key in ['numFailedTests','numPendingTests','numTodoTests','numFailedTestSuites','numPendingTestSuites']:assert unit[key]==0
 assert len(unit['testResults'])==2
 for suite in unit['testResults']:
  assert suite['status']=='passed' and suite['message']==''
  for case in suite['assertionResults']:assert case['status']=='passed' and case['failureMessages']==[]
 idx=rd('unit-green.json.reports-V7rVpm/index.json');assert idx['complete'] and idx['error']=='' and len(idx['entries'])==2
 good={'code':0,'noOutputTimedOut':False,'signal':None,'groupJoined':True};assert idx['merge']==good
 for entry,count in zip(idx['entries'],[15,35]):
  assert entry['state']=='finished' and entry['acceptedAttempt']==1 and len(entry['attempts'])==1 and entry['attempts'][0]['outcome']==good
  stem='unit-green.json.reports-V7rVpm/'+str(entry['invocation'])+'/1/report.json';report=rd(stem);cap=rd(stem+'.capture.json')
  assert entry['attempts'][0]['json'].endswith('/'+stem)
  assert report['success'] and report['numTotalTests']==report['numPassedTests']==count
  for key in ['numFailedTests','numPendingTests','numTodoTests','numFailedTestSuites','numPendingTestSuites']:assert report[key]==0
  assert cap['ended']=={'reason':'passed','unhandledErrors':0,'failedModules':0,'suiteErrors':0}
  assert not cap['processTimedOut'] and not cap['ignoreUnhandledErrors'] and not cap['passWithNoTests']
  assert len(cap['modules'])==1 and cap['modules'][0]['file'].endswith('/'+entry['includePatterns'][0])
 assert not re.search(r'\[test\]\s+retrying|no-output timeout|FATAL ERROR:|out of memory|unfinished|Unhandled Error',strip(text('unit-green.log.txt')),re.I)
 checks=strip(text('check-changed.log.txt'));assert re.findall(r'^\s*\S+\s+ok\s+(.+)$',checks,re.M)==PASSED
 assert checks.count('typescript(unbound-method)')==2 and ':248:31]' in checks and ':266:29]' in checks and 'Found 0 warnings and 2 errors.' in checks and '[check:changed] FAILED (exit 1)' in checks
 cases=rd('harness/cases.json');procs=rd('processes.json');assert len(cases)==len(procs['records'])==len(procs['inspections'])==12
 assert procs['quiescent'] and procs['ownedFixtureRemoved'] and procs['retainedState'] is None and procs['cancelledSignal'] is None
 assert len({p['pid'] for p in procs['records']}|{p['observerPid'] for p in procs['inspections']})==24
 results=[]
 for i,cell in enumerate(cases):
  stem=cell['id']+'/';p=rd(stem+'process.json');o=rd(stem+'inspection-normal.json');child=rd(stem+'child.json');screen=rd(stem+'screen.json');b=raw(stem+'raw.ansi');t=b.decode('utf-8')
  assert p==procs['records'][i] and o==procs['inspections'][i]
  assert all(p[k] is True for k in ['spawned','joined','eof']) and p['exitCode']==0 and not p['forced'] and p['groupState']=='dead' and 'error' not in p
  assert len(b)==p['rawBytes'] and sha(b)==p['rawSha256']
  assert all(o[k] is True for k in ['spawned','reaped','pipeEOF','kernelGroupAbsent']) and o['forcedSignals']==[] and o['exitCode']==0 and o['canonicalState']=='dead' and 'error' not in o
  obs=rd(stem+'inspection-normal.stdout');assert obs==o['observation'] and raw(stem+'inspection-normal.stderr')==b''
  assert obs['pid']==p['pid']==o['targetPid'] and obs['observerPid']==o['observerPid'] and obs['policy']=='indeterminate' and obs['canonicalState']=='dead' and obs['observationValid'] and obs['builtinBindingsRestored']
  assert len(obs['observations'])==1 and obs['observations'][0]['kind']=='kernel' and obs['observations'][0]['pid']==-p['pid'] and obs['observations'][0]['signal']==0 and obs['observations'][0]['error']['code']=='ESRCH'
  assert child['completed'] and child['listenersRestored'] and child['rawInputRestored'] and child['suppressed']==cell['expectSuppressed'] and child['repeatedStop']==cell.get('repeatStop',False)
  assert child['columns']==cell.get('finalColumns',cell['columns']) and child['initialColumns']==cell['columns']
  label=cell.get('label','Checking channel status (probe)…');updated=label.replace('Scanning','Checking') if 'label' in cell else 'Checking channel status (ready)…'
  assert re.findall(r'\x1b\]9;4;.*?\x1b\\',t,re.S)==['\x1b]9;4;3;;'+label+'\x1b\\','\x1b]9;4;3;;'+updated+'\x1b\\','\x1b]9;4;0;0;'+updated+'\x1b\\']
  phases=['begin']+(['resize-request'] if cell['kind']=='delayed' else [])+(['suppressed'] if child['suppressed'] else [])+['updated','active','done','settled'];assert re.findall(r'\x1b\]777;progress-proof;([^\x07]+)\x07',t)==phases and t.endswith('\x1b]777;progress-proof;settled\x07')
  pattern=r'\((?:\\/|\|\||--)\)' if cell['kind']=='wizard' else r'[◒◐◓◑•oO0]  '
  assert [r['byte'] for r in p['readySignals']]==['S','A']
  for n,r in enumerate(p['readySignals']):
   assert 0<r['offset']<=len(b);prefix=b[:r['offset']].decode('utf-8');count=len(re.findall(pattern,strip(prefix)));assert count==r['frames'];assert (count==0 if child['suppressed'] else count>=(1 if n==0 else 26))
   if n==1:assert '\x1b]777;progress-proof;updated\x07' in prefix
  active=t.index('\x1b]777;progress-proof;active\x07');done=t.index('\x1b]777;progress-proof;done\x07');completion=strip(t[active+len('\x1b]777;progress-proof;active\x07'):done]).replace('\r\n','\n').replace('\r','');expected=('' if 'finalMessage' not in cell else '◇  '+cell['finalMessage']+'\n') if cell['kind']=='wizard' else '◇  \n';assert completion==expected
  assert screen['done']==screen['settled'] and len([x for x in screen['active'] if x])==(0 if child['suppressed'] else 2)
  if cell['kind']=='delayed':assert strip(b[:p['resizeByteOffset']].decode())=='' and p['resizeApplied']
  if cell['id']=='wizard-unicode':assert '👨‍👩‍👧‍👦' in strip(t[:active]) and '…' in strip(t[:active])
  results.append(cell['id'])
 outcomes=rd('outcomes.json');assert outcomes['failures']==[] and [r['id'] for r in outcomes['results']]==results and all(r['conforms'] for r in outcomes['results'])
 assert sha(before)==repair['beforeWizardHash']=='4f50ae8ff9bde6f0acf727d6a9b58d6600c66f3d37acba6c335dd7c63d5fe458' and sha(after)==repair['afterWizardHash']=='9bdfeb7af6fc0aa8a2d6f3ebd763f01e73eec653d23f74a2b33fe78e4feec28e'
 expected=before.decode().replace('import { afterAll, afterEach, beforeEach, describe, expect, it, onTestFinished, vi } from "vitest";','import {\n  afterAll,\n  afterEach,\n  beforeEach,\n  describe,\n  expect,\n  it,\n  onTestFinished,\n  vi,\n  type MockInstance,\n} from "vitest";').replace('describe("createClackPrompter", () => {\n  beforeEach','describe("createClackPrompter", () => {\n  let write: MockInstance<typeof process.stdout.write>;\n\n  beforeEach').replace('    vi.spyOn(process.stdout, "write").mockImplementation(() => true);','    write = vi.spyOn(process.stdout, "write").mockImplementation(() => true);').replace('      const write = vi.mocked(process.stdout.write);\n','').replace('    const write = vi.mocked(process.stdout.write);\n','')
 assert after.decode()==expected
 return {'accepted':True,'originalRun':lineage['run'],'originalJobConclusion':'failure','originalWizardHash':repair['beforeWizardHash'],'candidateWizardHash':repair['afterWizardHash'],'unitPasses':50,'ptyCases':results,'naturalProcessReceipts':24,'staticPassed':PASSED,'remaining':'fresh lint, eight later static checks, final guards','equivalence':'same existing beforeEach spy handle and unchanged assertions; old execution remains attributed to old test hash'}

def load(lane):
 lineage=json.loads((lane/'REUSE-LINEAGE.json').read_text());data={f:(lane/'reuse'/f).read_bytes() for f in lineage['files']};repair=json.loads((lane/'REPAIR.json').read_text());return data,lineage,repair,(lane/'before.test.ts').read_bytes(),(lane/'after.test.ts').read_bytes()
if __name__=='__main__':print(json.dumps(validate(*load(Path(sys.argv[1])))))
